buffer/src/main/java/io/netty/buffer/MiMallocByteBufAllocator.java

```diff
@@ -0,0 +1,2377 @@
+        // `use_delayed` will only be true if `threadDelayedFreeFlag == USE_DELAYED_FREE`.
+        if (use_delayed) {
+            // Racy read on `heap`, but ok because `DELAYED_FREEING` is set.
+            // (see `heapCollectAbandon`)
+            LocalHeap heap = page.heapx.get();
+            if (heap != null) {
+                // Add to the delayed free list of this heap.
+                Block dfree;
+                do {
+                    dfree = heap.threadDelayedFreeList.get();
+                    block.nextBlock = dfree;
+                } while (!heap.threadDelayedFreeList.compareAndSet(dfree, block));
+            }
+            // Reset the `DELAYED_FREEING` flag.
+            boolean isReset;
+            do {
+                isReset = page.threadDelayedFreeFlag.compareAndSet(DELAYED_FREEING, NO_DELAYED_FREE);
+            } while (!isReset);
+        } else { // Common path
+            Block current;
+            do {
+                current = page.threadFreeList.get();
+                block.nextBlock = current;
+            } while (!page.threadFreeList.compareAndSet(current, block));
+        }
+    }
+
+    // Free huge block from another thread
+    private void segmentHugePageFree(Segment segment, Page page, Block block) {
+        // Huge page segments are always abandoned and can be freed immediately by any thread claim it and free.
+        LocalHeap heap = THREAD_LOCAL_HEAP.get();
+        // If this is the last reference, the CAS should always succeed.
+        if (segment.ownerThread.compareAndSet(null, Thread.currentThread())) {
+            block.nextBlock = page.freeList;
+            page.freeList = block;
+            page.usedBlocks--;
+            heap.segmentPageFree(page, true);
+        }
+    }
+
+    private void freeGenericMt(Page page, Segment segment, Block block) {
+        freeBlockMt(page, segment, block);
+    }
+
+    private void freeGenericLocal(Page page, Segment segment, Block block, LocalHeap heap) {
+        page.freeBlockLocal(block, true, heap);
+    }
+
+    /**
+     * Allocate into the given buffer. Used by {@link AdaptivePoolingAllocator.AdaptiveByteBuf#capacity(int)}.
+     */
+    void reallocate(int size, int maxCapacity, MiByteBuf into) {
+        MiByteBuf result = allocate(size, maxCapacity, into);
+        assert result == into: "Re-allocation created separate buffer instance";
+    }
+
+    ByteBuf allocate(int size, int maxCapacity) {
+        return allocate(size, maxCapacity, null);
+    }
+
+    private MiByteBuf allocate(int size, int maxCapacity, MiByteBuf byteBuf) {
+        LocalHeap localHeap = THREAD_LOCAL_HEAP.get();
+        int wSize = toWordSize(size);
+        if (size <= PAGES_FREE_DIRECT_SIZE_MAX) {
+            Page page = localHeap.pagesFreeDirect[wSize];
+            // Fast path
+            Block block = page.freeList;
+            if (block != null) {
+                if (byteBuf == null) {
+                    byteBuf = block;
```

Comment:
In the best case scenario the pooling of buffers is not using a separate storage but is both of the right size (class) and the wrapper too.